Redesign images overview

jelly opened the pull request:
Part of the cockpit-podman redesign https://github.com/cockpit-project/cockpit/discussions/16059

KKoukiou reviewed (approved):
The change is straightforward I just have one comment. The same logic exist for VMs as well. In that case the dialog has a checkbox, 'Immediately start VM' instead of 'Start VM after creation' which is the equivalent of the text chosen here. I feel it's nice to keep the texts same, what do you think @garrett? 

No vetoing - I just want to make it a conscious decision.

jelly commented:
> The change is straightforward I just have one comment. The same logic exist for VMs as well. In that case the dialog has a checkbox, 'Immediately start VM' instead of 'Start VM after creation' which is the equivalent of the text chosen here. I feel it's nice to keep the texts same, what do you think @garrett?
> 
> No vetoing - I just want to make it a conscious decision.
> 
> ![Screen Shot 2021-08-31 at 09 29 24](https://user-images.githubusercontent.com/14921356/131461759-63fb5d3d-986b-4bf2-865e-3c1090bfa276.png)

There was a discussion about this [here](https://github.com/cockpit-project/cockpit/discussions/16059#discussioncomment-1194000), so that's how I came up with the text.

I am however not sure if it should be the default :thinking:

jelly commented on src/ImageRunModal.jsx:
```diff
@@ -237,9 +237,10 @@ export class ImageRunModal extends React.Component {
             memoryUnit: 'MiB',
```
Should I rename the whole dialog/file from `run` to `create`?

marusak replied:
You might, not necessary though

jelly replied:
I'll leave it as todo for later, as I feel that the PR is already quite big.

marusak commented on src/ImageRunModal.jsx:
```diff
@@ -417,17 +423,20 @@ export class ImageRunModal extends React.Component {
                    position="top" variant="medium"
                    onClose={this.props.close}
-                   title={_("Run image")}
+                   title={_("Create image")}
```
Not create "image" but "container".

marusak commented:
No option about "Immediately start" vs "Start after creation".  We can stay consistent with other pages for sure. Or just try to keep the string as short as possible.

jelly commented:
The last test failures are due to the `tags` missing from the `image details` overview, is this is something missing from the redesign @marusak ?

jelly commented:
Updated images overview page

![image](https://user-images.githubusercontent.com/67428/131667768-fa1658b5-3c67-4feb-a5a3-54b58a3b08cf.png)

Updated "create container" modal

![image](https://user-images.githubusercontent.com/67428/131667797-47cccc4b-23c5-4c2c-8cbf-87cca13cd232.png)

marusak commented:
Notice, that when one image has more than one tag, cmdline shows it as two separate lines. (the fedora ones)

![Screenshot from 2021-09-01 14-13-01secpmd](https://user-images.githubusercontent.com/12330670/131669224-a57c01e7-9d7c-47c7-8312-fe64f4c5ad88.png)

You show it as one line with two names. That seems odd. It should be two lines as well.

![Screenshot from 2021-09-01 14-11-38](https://user-images.githubusercontent.com/12330670/131669174-baa70918-27d5-4110-adde-f3e797e491a1.png)

marusak commented:
Also the common question - do we want MiB or MB units? :) I would say, that MB are better choice in this situation + is consistent with podman cmd.

martinpitt commented:
> I would say, that MB are better choice in this situation

:+1 -- IMHO, base-2 makes very little sense these days, pretty much only on RAM size.

jelly commented:
> Notice, that when one image has more than one tag, cmdline shows it as two separate lines. (the fedora ones)
> 
> ![Screenshot from 2021-09-01 14-13-01secpmd](https://user-images.githubusercontent.com/12330670/131669224-a57c01e7-9d7c-47c7-8312-fe64f4c5ad88.png)
> 
> You show it as one line with two names. That seems odd. It should be two lines as well.
> 
> ![Screenshot from 2021-09-01 14-11-38](https://user-images.githubusercontent.com/12330670/131669174-baa70918-27d5-4110-adde-f3e797e491a1.png)

We can change it to be consistent to with podman output, I do wonder if the column name is then maybe confusing as it says `Image`.

KKoukiou commented on src/Images.jsx:
```diff
@@ -60,14 +60,31 @@ class Images extends React.Component {
 
         const columns = [
             { title: image.RepoTags ? image.RepoTags[0] : "<none>:<none>", header: true },
+            { title: image.isSystem ? _("system") : <><span className="images-grey-text">user: </span>{this.props.user}</> },
```
This needs to be translated. _("user: ")

KKoukiou replied:
Also this breaks in small screens.
![Screen Shot 2021-09-01 at 17 45 46](https://user-images.githubusercontent.com/14921356/131702597-a48c8697-0aac-4ba9-a382-fd040e3254bb.png)

martinpitt replied:
Please avoid translating trailing (or leading) spaces -- it is pretty much garanteed to get messed up during the translation process. Use CSS or HTML structure to do spacing.

jelly replied:
Added translations without a trailing space, for the small screens I've put the <span> with the {this.props.user} in a <div> so it doesn't break as it does now. I'm not sure if this is the best solution.

KKoukiou commented on src/Images.css:
```diff
@@ -16,3 +16,11 @@
+}
+
+.images-grey-text {
+   color: var(--pf-global--Color--200);
```
Whoa :) Thanks for using PF colors!

KKoukiou commented on src/Images.jsx:
```diff
@@ -60,14 +60,31 @@ class Images extends React.Component {
+        }
+        const containers = imageContainerList[image.Id + image.isSystem.toString()];
+        if (containers !== undefined) {
+            const fmt = containers.length === 1 ? _("$0 container") : _("$0 containers");
```
Better done like this:
const fmt = cockpit.format(cockpit.ngettext("$0 container", "$0 containers", containers.length), containers.length)

jelly replied:
Oh nice! I should have checked https://cockpit-project.org/guide/latest/cockpit-locale.html I guess!

KKoukiou commented on src/ImageDetails.jsx:
```diff
@@ -9,37 +9,31 @@ const _ = cockpit.gettext;
-            <DescriptionListGroup>
-                <DescriptionListTerm>{_("Tags")}</DescriptionListTerm>
-                <DescriptionListDescription>{image.RepoTags ? image.RepoTags.join(" ") : ""}</DescriptionListDescription>
+                <DescriptionListTerm>{_("Default command")}</DescriptionListTerm>
```
I know you copied it from the design document - but I feel like Default is not needed here. I would just keep it Command. What do you think @marusak ?

marusak replied:
https://docs.docker.com/engine/reference/builder/#cmd

:man_shrugging: It is kind of default command, but also not super important to spell out "default"...

KKoukiou commented on src/ImageRunModal.jsx:
```diff
@@ -417,17 +423,20 @@ export class ImageRunModal extends React.Component {
                                      default={{ envKey: null, envValue: null }}
                                      itemcomponent={ <EnvVar />} />
                 </FormGroup>
+                <FormGroup fieldId='run-image-dialog-run' label={_("Start after creation")}>
```
The fieldId has to be the same with the id of the field that is nested in the FormGroup. So in this case 'run'.
Please use more complicated ids however, not 'run' as in one page ids should be unique. and single word string tend to become not unique :)

KKoukiou commented on src/Images.jsx:
```diff
@@ -202,10 +232,27 @@ class Images extends React.Component {
+        const imageTitleStats = (
+            <div className="images-title-stats">
+                <Text component={TextVariants.h5}>
+                    {cockpit.format(_("$0 images total $1"), imageStats.imagesTotal, cockpit.format_bytes(imageStats.imagesSize))}
```
The design doc has a comma after total.

KKoukiou commented on src/Images.jsx:
```diff
@@ -202,10 +232,27 @@ class Images extends React.Component {
+                </Text>
+                {imageStats.unusedTotal !== 0 &&
+                <Text component={TextVariants.h5}>
+                    {cockpit.format(_("$0 unused images $1"), imageStats.unusedTotal, cockpit.format_bytes(imageStats.unusedSize))}
```
The design doc has a comma after 'images'

KKoukiou reviewed (changes requested):
First quick pass for today - I will do a real one tomorrow :) Pretty amazing work! Thanks for using PF4 components and keeping the commit simple to review.

KKoukiou commented on src/Images.css:
```diff
@@ -16,3 +16,24 @@
+   color: var(--pf-global--Color--200);
+}
+
+.images-title {
```
How about wrapping the title with the stats in a https://www.patternfly.org/v4/layouts/flex/#basic. 

That wil add autopadding as well.

jelly replied:
Thanks, that works nicely and that means I can remove all the CSS yay!

marusak commented:
> We can change it to be consistent to with podman output, I do wonder if the column name is then maybe confusing as it says Image.

We don't want column 'tag' so repository would not do. It is repo + image name + tag. I think `Image` is then fine, it is obvious whats in the column anyway :)

KKoukiou commented on src/Images.jsx:
```diff
@@ -202,10 +232,29 @@ class Images extends React.Component {
+        const imageTitleStats = (
+            <>
+                <Text component={TextVariants.h5}>
+                    {cockpit.format(_("$0 images total, $1"), imageStats.imagesTotal, cockpit.format_bytes(imageStats.imagesSize))}
```
This would need ngettext - in case there is one image.

KKoukiou commented on src/Images.jsx:
```diff
@@ -202,10 +232,29 @@ class Images extends React.Component {
+                </Text>
+                {imageStats.unusedTotal !== 0 &&
+                <Text component={TextVariants.h5}>
+                    {cockpit.format(_("$0 unused images, $1"), imageStats.unusedTotal, cockpit.format_bytes(imageStats.unusedSize))}
```
This would need ngettext - in case there is one image.

KKoukiou commented on src/Images.css:
```diff
@@ -16,3 +16,11 @@
     --pf-c-table__action--PaddingBottom: 0.5rem;
 }
+
+#containers-images .pf-c-table.pf-m-compact .pf-c-table__toggle {
```
This will create inconsistent left padding between the images and containers toggles in lists. 
Could you please make it apply also for Containers and move it to podman.scss file?

KKoukiou replied:
Something like this I mean:

``` diff
diff --git src/Images.css src/Images.css
index 3caed167..f080943f 100644
--- src/Images.css
+++ src/Images.css
@@ -17,10 +17,6 @@
     --pf-c-table__action--PaddingBottom: 0.5rem;
 }
 
-#containers-images .pf-c-table.pf-m-compact .pf-c-table__toggle {
-    padding-left: 0px;
-}
-
 .images-grey-text {
    color: var(--pf-global--Color--200);
 }
diff --git src/podman.scss src/podman.scss
index 43aa363f..d1ad85af 100644
--- src/podman.scss
+++ src/podman.scss
@@ -8,6 +8,12 @@
     @extend .ct-card;
 }
 
+#containers-images, #containers-containers {
+    .pf-c-table.pf-m-compact .pf-c-table__toggle {
+        padding-left: 0px;
+    }
+}
+
 .container-block {
     display: flex;
     flex-direction: column;

```

KKoukiou replied:
And last thing - because we are all not really good with CSS I like adding comment so that we don't get lost. 
In this case, decrease default padding for the toggle button in list.

jelly commented:
> Also the common question - do we want MiB or MB units? :) I would say, that MB are better choice in this situation + is consistent with podman cmd.

switched to MB everywhere.

marusak commented on src/ImageRunModal.jsx:
```diff
@@ -417,17 +423,20 @@ export class ImageRunModal extends React.Component {
                                      default={{ envKey: null, envValue: null }}
                                      itemcomponent={ <EnvVar />} />
                 </FormGroup>
+                <FormGroup fieldId='run-image-dialog-start-after-creation' label={_("Start after creation")}>
```
![Screenshot from 2021-09-03 09-10-15](https://user-images.githubusercontent.com/12330670/131965300-dc367daf-31f6-49df-a988-11fb6aa29e99.png)

It seems misaligned with the label.  If you add `hasNoPaddingTop` prop to this `FormGroup` it should work as expected.

marusak commented on test/check-application:
```diff
@@ -1334,7 +1334,7 @@ class TestApplication(testlib.MachineCase):
         b.click('#containers-containers tbody tr:contains("busybox:latest") td.pf-c-table__toggle button')
         b.click('#containers-images tbody tr:contains("busybox:latest") .ct-container-create')
-        b.wait_visible('div.pf-c-modal-box header:contains("Run image")')
+        b.wait_visible('div.pf-c-modal-box header:contains("Create container")')
```
Can you also test that new checkbox - that you will unset it and make sure it does not start?

jelly replied:
Added a test.

marusak reviewed (changes requested):
The new checkbox is misaligned and a test would be nice :) Otherwise looks great!

marusak commented on test/check-application:
```diff
@@ -1425,6 +1425,24 @@ class TestApplication(testlib.MachineCase):
+        b.click('.pf-c-modal-box__footer button:contains("Create")')
+        b.wait_not_present("div.pf-c-modal-box")
+
+        b.wait_not_in_text("#containers-containers", container_name)
```
This is race-y. It waits for container not to be in the list - which might be just true by default (tests are fast so it might not yet be present), then changes to all containers and then waits for it to be present. It unfortunately does not say anything whether the container is running or not.

It should be safe to just switch to `all` and wait for the container status to be `created`.

jelly replied:
Looking at the test of the tests, it might be better to use `waitContainer`

marusak reviewed (approved):
Awesome work, thank you!

garrett commented:
This is awesome to see! :+1: :partying_face: 

Minor nitpick: We should never have checkboxes without labels. They're inaccessible and the click target is much smaller (as the label itself is part of the click target when done in a proper, accessible way) and that benefits everyone, even people with full sight using a mouse.

Also, I agree we should be consistent where possible with language.
